### ISSUE TITLE
Weird behavior with custom `algorithm_with_properties<>`

### ISSUE THREAD

**Issue by Christian-Stieber** [boostorg/fiber#308]
I've been trying to hack up a way to wait for all fibers to complete, and I always end up with a count of 2 instead of 1 at the end. Here's a full example -- sorry, it's a couple of lines.

```
#include <boost/fiber/properties.hpp>
#include <boost/fiber/algo/algorithm.hpp>
#include <boost/fiber/algo/round_robin.hpp>
#include <boost/fiber/mutex.hpp>
#include <boost/fiber/condition_variable.hpp>
#include <boost/version.hpp>

#include <thread>
#include <iostream>

/************************************************************************/

class Scheduler;

/************************************************************************/

class Properties : public boost::fibers::fiber_properties
{
private:
    Scheduler* scheduler=nullptr;

public:
    Properties(boost::fibers::context* context)
        : fiber_properties(context)
    {
        assert(false);
    }

    Properties(boost::fibers::context*, Scheduler&);
    ~Properties();
};

/************************************************************************/

class Scheduler : public boost::fibers::algo::algorithm_with_properties<Properties>
{
public:
    boost::fibers::mutex mutex;
    boost::fibers::condition_variable condition;
    unsigned int counter=0;

    boost::fibers::algo::round_robin scheduler;

public:
    Scheduler(Scheduler*& self)
    {
        self=this;
    }

public:
    virtual void awakened(boost::fibers::context* context, Properties&) noexcept // override
    {
        return scheduler.awakened(context);
    }

    virtual boost::fibers::context* pick_next() noexcept override
    {
        return scheduler.pick_next();
    }

    virtual bool has_ready_fibers() const noexcept override
    {
        return scheduler.has_ready_fibers();
    }

    virtual void suspend_until(std::chrono::steady_clock::time_point const& when) noexcept override
    {
        return scheduler.suspend_until(when);
    }

    virtual void notify() noexcept override
    {
        return scheduler.notify();
    }

public:
    virtual boost::fibers::fiber_properties* new_properties(boost::fibers::context* context) override
    {
        return new Properties(context, *this);
    }

    void wait()
    {
        std::unique_lock<decltype(mutex)> lock(mutex);
        condition.wait(lock, [this](){ return counter==1; });
    }
};

/************************************************************************/

Properties::Properties(boost::fibers::context* context, Scheduler& scheduler_)
    : fiber_properties(context), scheduler(&scheduler_)
{
    scheduler->counter++;
    std::cout << "Fiber count: " << scheduler->counter << std::endl;
}

/************************************************************************/

Properties::~Properties()
{
    scheduler->counter--;
    std::cout << "Fiber count: " << scheduler->counter << std::endl;
    scheduler->condition.notify_one();
}

/************************************************************************/

int main()
{
    std::cout << "BOOST_VERSION: " << BOOST_VERSION << std::endl;

    std::thread([](){
        Scheduler* scheduler=nullptr;
        boost::fibers::use_scheduling_algorithm<Scheduler>(scheduler);
        for (int i=0; i<10; i++)
        {
            boost::fibers::fiber([i](){
                boost::this_fiber::sleep_for(std::chrono::seconds(2+i));
            }).detach();
        }
        std::cout << "Waiting..." << std::endl;
        scheduler->wait();
        std::cout << "All done!" << std::endl;
    }).join();
}
```

This is on a Debian "testing", to get a halfway-recent g++ and boost; I'm running this as:

```
stieber@debian:~$ g++ -ggdb -std=c++20 Test.cpp -l boost_fiber -l boost_context && ./a.out
BOOST_VERSION: 108100
Fiber count: 1
Fiber count: 2
Fiber count: 3
Fiber count: 4
Fiber count: 5
Fiber count: 6
Fiber count: 7
Fiber count: 8
Fiber count: 9
Fiber count: 10
Fiber count: 11
Waiting...
Fiber count: 10
Fiber count: 11
Fiber count: 10
Fiber count: 9
Fiber count: 8
Fiber count: 7
Fiber count: 6
Fiber count: 5
Fiber count: 4
Fiber count: 3
Fiber count: 2
```

Up until "waiting", all is expected -- I launch 10 fibers, and I have my initial fiber, for a total of 11.

Things start getting weird right away, though: I have no explanation for "10 11 10", but if we assume that "somehow" a new property instance is created, then the ending makes sense: it stops at 2, and never goes down to 1.

Also, if I change the `wait()` call to wait for 2 instead of 1, it counts down to 0, then segfaults. Maybe the "fake" property that showed up, getting cleaned up after the scheduler or something like that. I'll probably just try to wait for the 2 and replace the `Scheduler*` with a `weak_ptr` construct for now.
